Initial perf jitdump implementation

sdmaclea opened the pull request:
This initial commit:

- creates the jit-%d.dump file
- writes the file header
- creates the mmap mark
- adds `JIT_CODE_LOAD` records for JIT generated code and for stubs.

It has been lightly tested on Ubuntu 16.04 with perf built from Ubuntu-azure-4.15.0-1057.62 with the patch https://lkml.org/lkml/2019/9/20/862 

```bash
export COMPlus_PerfMapEnabled=1
~/git/linux-azure-xenial/tools/perf/perf record -k 1 -g /home/stmaclea/git/coreclr/bin/tests/Linux.x64.Debug/Tests/Core_Root/corerun helloworld.dll
~/git/linux-azure-xenial/tools/perf/perf --debug verbose=10 inject --input perf.data --jit --output perfjit.data
~/git/linux-azure-xenial/tools/perf/perf report --input perfjit.data
```

Basic functionality seems correct, however there are sample which are hitting in `perf-*.map`, but not found in `jit-*.dump`.

It needs to be further debugged.

Fixes #26842 

/cc @dotnet/dotnet-diag

sdmaclea committed:
Initial perf jitdump implementation

sdmaclea commented:
I looked at the missing symbols.  It turns out when we grow the code page, we remap the entire page. This obscures the previous `JIT_CODE_LOAD` records. Emitting a `JIT_CODE_MOVE` for all previous `JIT_CODE_LOAD` records will likely solve this problem.

sdmaclea committed:
Fix spelling

sdmaclea committed:
Correct JIT_DUMP_MAGIC

Writer should always write the same value, independent of endianness.

janvorli commented on src/pal/src/misc/perfjitdump.cpp:
```diff
@@ -0,0 +1,271 @@
+
+        snprintf(jitdumpPath, sizeof(jitdumpPath), "/tmp/jit-%i.dump", getpid());
+
+        fd = open(jitdumpPath, O_CREAT|O_TRUNC|O_RDWR|O_CLOEXEC, S_IRUSR|S_IWUSR );
```
We need to add checks for failures at many places (open, write, fsync, ...)

janvorli replied:
We should not assume that the TMP folder is at /tmp. We should try to read TMPDIR envvar first and use /tmp only if that variable is not set. Like we do in GetTempPathA.

janvorli commented on src/inc/perfjitdump.h:
```diff
@@ -0,0 +1,29 @@
+#ifndef PERF_JITDUMP_H
+#define PERF_JITDUMP_H
+
+struct PerfJitDumpState;
```
All things implemented in PAL and exposed for the runtime are in pal.h are just plain C functions. We should do the same here. That means expose something like PAL_PerfJitDumpStart, PAL_PerfJitDumpLogMethod and PAL_PerfJitDumpFinish.

noahfalk commented on src/vm/perfmap.cpp:
```diff
@@ -50,6 +51,9 @@ void PerfMap::Initialize()
         {
             s_ShowOptimizationTiers = true;
         }
+#ifndef CROSSGEN_COMPILE
```
Nit: Have JITDUMP_SUPPORTED set to false in CROSSGEN builds so we don't need to surround every API call with these ifdefs?

noahfalk reviewed (commented):
Looks fine so far, aside from the stuff Jan already noted

sdmaclea committed:
Add PreJIT to jitdump

sdmaclea commented:
**Status:**

The draft functional implementation is complete.

Testing with perf built from:
```
Repo: git://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git 
Branch:perf/core
Patches:
- https://lkml.org/lkml/2019/9/20/862
- https://lkml.org/lkml/2019/10/2/875

cd tools/perf && make DEBUG=1
```

I captured a fairly simple app that printed hello world + a counter 100K times.

I was able to see and annotate R2R and JIT generated methods some for multiple compilation tiers.

The vast majority of symbols were resolved.  There were a handful of unknown addresses.  I think we have missed logging of one type of stub.

The call stacks looked believable and complete. Unwindinfo didn't seem critically important for 'fp' based perf callstacks. They are probably needed for perf dwarf call stack support.

**ToDo:**

- Review comments
  - Clean up PAL interface
  - Clean up #ifdef
  - Add some basic error checking
- Add configuration knobs
  - jitdump dirpath
  - disable jitdump separately?

**Future**

- Add unwind info.
- Add debug info.

sdmaclea committed:
Feedback

Revise PAL API
Add error handling
Clean up #ifdef CROSSGEN
Add COMPlus_PerfMapJitDumpPath
Add placeholder debuginfo and unwindinfo arguments

sdmaclea commented:
I have addressed all feedback and finished my to do list. This should be ready for review.

/cc @brianrob

jkotas commented on src/inc/clrconfigvalues.h:
```diff
@@ -562,6 +562,7 @@ RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_ProfAPI_ValidateNGENInstrumentation, W("Pro
 
 #ifdef FEATURE_PERFMAP
 RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_PerfMapEnabled, W("PerfMapEnabled"), 0, "This flag is used on Linux to enable writing /tmp/perf-$pid.map. It is disabled by default", CLRConfig::REGUTIL_default)
+CONFIG_STRING_INFO_EX(EXTERNAL_PerfMapJitDumpPath, W("PerfMapJitDumpPath"), "Specifies a path to write the perf jitdump file. Defaults to GetTempPathA()", CLRConfig::REGUTIL_default)
```
Should this be `RETAIL_CONFIG_DWORD_INFO_EX` so that this is available even in the release builds?

sdmaclea committed:
RETAIL ... PerfMapJitDumpPath

janvorli commented on src/vm/perfmap.cpp:
```diff
@@ -51,9 +51,20 @@ void PerfMap::Initialize()
+        // CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_PerfMapJitDumpPath) returns a LPWSTR
+        // Use GetEnvironmentVariableA because it is simpler.
+        // Keep comment here to make it searchable.
+        auto len = GetEnvironmentVariableA("COMPlus_PerfMapJitDumpPath", jitdumpPath, sizeof(jitdumpPath) - 1);
```
A nit - can you please use a specific type instead of `auto`? The convention is to use auto only in places where the type cannot be expressed otherwise, like for lambdas.

janvorli commented on src/vm/perfmap.cpp:
```diff
@@ -274,6 +281,11 @@ void PerfMap::LogPreCompiledMethod(MethodDesc * pMethod, PCODE pCode)
 {
     LIMITED_METHOD_CONTRACT;
 
+    if (s_Current == nullptr)
```
It seems that if one of these logging functions races with the PerfMap::Destroy, we have a potential to crash in the code below.

sdmaclea replied:
This doesn't look like an issue with this function (since it is just an early out), but for other preexisting functions it is potentially an issue.

I'll open a separate PR to fix this.

sdmaclea committed:
Feedback

Revise error handling
Remove perfjitdump.h
Fix #if CROSSGEN_COMPILE

AndyAyersMS commented:
@sdmaclea does this work out of the box or do I need some special build?

sdmaclea commented:
To fix all known issues, you should use a build of perf from the linux tip.

The last known issue was fixed with https://github.com/torvalds/linux/commit/c8f6ae1fb28d74ef7a0966477f61d1c2c8ee2e1e

Building perf is not that hard. 
+ Install kernel required tool chains (See for instance https://wiki.ubuntu.com/Kernel/BuildYourOwnKernel Build Environment section)
+ cd tools\perf
+ make

brianrob commented:
Also worth noting that `perfcollect` now has support for using this functionality by default.  FWIW, I have had reasonable luck using versions of `perf` that are available with most recent distros, but if you are running into issues, building the tip is a good way to ensure that they are not known issues that @sdmaclea has already fixed.

AndyAyersMS commented:
Ok, thanks... will give this a try.

AndyAyersMS commented:
Confessing my ignorance here. What can I do If I'm stuck on some older kernel version ( say 4.10)... does building the latest perf from there work? I end up with `perf version 4.4.232`.

I also realized I needed to run `perf inject` to pick up the jit events. 

Now I can sometimes see jitted code, eg one of the five methods below:
```
+   90.36%     0.00%  corerun          libcoreclr.so               [.] RunMain                                                                                                                                                              ▒
+   89.63%     0.02%  corerun          perf-36392.map              [.] void [r41741] X::Main()[Optimized]                                                                                                                                   ▒
+   64.42%     3.16%  corerun          perf-36392.map              [.] void [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::IntroSort(valuetype System.Span`1<!0>,int32)[OptimizedTier1]        ▒
+   63.75%     0.05%  corerun          jitted-36392-1185.so        [.] instance void [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::Sort(valuetype System.Span`1<!0>,class System.Collections.G▒
+   33.08%    33.06%  corerun          perf-36392.map              [.] void [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::InsertionSort(valuetype System.Span`1<!0>)[OptimizedTier1]          ▒
+   27.36%    27.35%  corerun          perf-36392.map              [.] int32 [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::PickPivotAndPartition(valuetype System.Span`1<!0>)[OptimizedTier1] ▒
+   16.63%     0.00%  corerun          libcoreclr.so               [.] MethodDesc::DoPrestub 
```
Do the jit events this play well with the tiering annotations? I notice above you were running debug so didn't see tiering.

Also `COMPlus_PerfMapShowOptimizationTiers=0` does not seem to suppress these annotations.

brianrob commented:
Yes, building a new perf and using it with an older kernel should be just fine.  For the most part, the user-mode tool is backwards compatible, and I've been told forwards compatible...

Yes, @kouvel did some work a while back to add tiering labels to the symbols so they should show up just fine in this view.  In PerfView they are stripped but can be enabled by looking at an optimization tiers view (you'll see it in the view name) or running PerfView with /ShowOptimizationTiers to enable it for all stack views.

AndyAyersMS commented:
Thanks all. Building the latest seems to be working:
```
+   89.74%     0.01%  corerun          jitted-44217-820.so   [.] void [r41741] X::Main()
+   64.11%     3.21%  corerun          jitted-44217-1150.so  [.] void [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::IntroSort(valuetype System.Span`1<!0>,int32)
+   63.38%     0.03%  corerun          jitted-44217-1185.so  [.] instance void [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::Sort(valuetype System.Span`1<!0>,class System.Collections.Generic.
+   33.08%    33.08%  corerun          jitted-44217-1165.so  [.] void [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::InsertionSort(valuetype System.Span`1<!0>)
+   27.14%    27.14%  corerun          jitted-44217-1164.so  [.] int32 [System.Private.CoreLib] System.Collections.Generic.GenericArraySortHelper`1[System.Int32]::PickPivotAndPartition(valuetype System.Span`1<!0>)
+   16.78%     0.01%  corerun          libcoreclr.so         [.] MethodDesc::DoPrestub
```
